extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/OtelWebSocketEndpoint.java

```diff
@@ -0,0 +1,92 @@
+package io.quarkus.websockets.next.runtime.telemetry;
+
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_CLIENT_ATTR_KEY;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_CLOSED;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_ENDPOINT_ATTR_KEY;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_ID_ATTR_KEY;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_OPENED;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.URI_ATTR_KEY;
+
+import java.util.function.Function;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.Tracer;
+import io.quarkus.websockets.next.WebSocketClientConnection;
+import io.quarkus.websockets.next.WebSocketConnection;
+import io.quarkus.websockets.next.runtime.WebSocketEndpoint;
+import io.vertx.core.Future;
+
+/**
+ * {@link WebSocketEndpoint} wrapper that produces OpenTelemetry spans for opened and closed connection.
+ */
+final class OtelWebSocketEndpoint extends AbstractWebSocketEndpointWrapper {
+
+    /**
+     * Target ID represents either endpoint id or client id.
+     */
+    private final String targetIdKey;
+    private final String targetIdValue;
+    /**
+     * Span context for an HTTP request used to establish the WebSocket connection.
+     */
+    private final SpanContext initReqSpanContext;
+    private final Tracer tracer;
+    private final String connectionId;
+    private final String path;
+    private volatile SpanContext onOpenSpanContext = null;
+
+    OtelWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketConnection connection,
+            SpanContext initReqSpanContext, String path) {
+        super(delegate);
+        this.tracer = tracer;
+        this.initReqSpanContext = initReqSpanContext;
+        this.connectionId = connection.id();
+        this.targetIdKey = CONNECTION_ENDPOINT_ATTR_KEY;
+        this.targetIdValue = connection.endpointId();
+        this.path = path;
+    }
+
+    OtelWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketClientConnection connection, String path) {
+        super(delegate);
+        this.tracer = tracer;
+        this.initReqSpanContext = null;
+        this.connectionId = connection.id();
+        this.targetIdKey = CONNECTION_CLIENT_ATTR_KEY;
+        this.targetIdValue = connection.clientId();
+        this.path = path;
+    }
+
+    @Override
+    public Future<Void> onOpen() {
+        return delegate.onOpen().map(new Function<Void, Void>() {
+            @Override
+            public Void apply(Void unused) {
+                var span = createSpan(CONNECTION_OPENED, initReqSpanContext);
+                OtelWebSocketEndpoint.this.onOpenSpanContext = span.getSpanContext();
+                span.end();
```

Comment:
You can give me a hint what would such measurement mean? Is it time between initial request and finished HTTP upgrade and or something else? I'd expect that this is internal Quarkus logic that takes couple of ms. Anywa, yes we can if you want to, before HTTP upgrade is performed, I take span context from the initial request, so we can also create instant there and then, make different with instant here. Or am I missing something?

Reply:
TL;DR; I think it can be done, but I didn't see it in original plan and I'd need more details. We can chat together about it later as you suggest, it would be easier for me to understand.

Reply:
Starting before the HTTP upgrade is performed makes sense and is more useful to spot connection issues.

Reply:
But before the HTTP upgrade starts or before even WS connection is opened and the endpoint is called, you don't know if the upgrade will happen. Technically it is not a problem for me to move it there, but I'd like to check it is really what you want. They are traces for open connections, but before the HTTP Upgrade, connection is opened.

Other thing we can do is to measure different between initial http request and opened WebSockets exception if that makes sense. I'll probably ping you because this format is bit clumsy to clarify what is required.

Reply:
I think this one is done, I've moved where connection opening is tracked and added on connection opening error.